### PR TITLE
update install.sh

### PR DESCRIPTION
**Pull request by nikandfor**
Checked the script by shellcheck.

Changed chmod; mv to install. chmod can fail if +x mode is disabled for tmp directory.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -59,19 +57,21 @@ cleanup() {
...
   exit 1
 fi
 
-tar xzf "$TEMP_FILE" -C "$TEMP_FOLDER"
+tar xzf "$TEMP_FILE" -C "$TEMP_FOLDER" codecrafters
 
-chmod 0755 "$TEMP_FOLDER/codecrafters"
-if ! mv "$TEMP_FOLDER/codecrafters" "$INSTALL_PATH" 2> /dev/null; then
-  sudo -k mv "$TEMP_FOLDER/codecrafters" "$INSTALL_PATH"
+if ! install "$TEMP_FOLDER/codecrafters" "$INSTALL_PATH" 2>/dev/null; then
```
Comment (rohitpaulk) [inline install.sh:71]: @nikandfor how sure are we that `install` is a good option re: compatibility? It looks like popular scripts don't use `install` and use `chmod`/`mv` instead, example:

- https://sentry.io/get-cli/
- https://fly.io/install.sh
- https://deno.land/x/install/install.sh

File: install.sh [hunk 2]
```diff
@@ -59,19 +57,21 @@ cleanup() {
 }
 
 trap cleanup EXIT
+
+echo Downloading the archive...
```
Comment (rohitpaulk) [inline install.sh:61]: This extra log line can be removed, curl is configured below to emit a progress bar, so it'll be clear that something is going on

File: install.sh [hunk 3]
```diff
@@ -7,45 +7,43 @@ VERSION=${SENTRY_CLI_VERSION:-v19}
...
+esac
+
+INSTALL_DIR=${INSTALL_DIR:-/usr/local/bin}
+INSTALL_PATH=${INSTALL_PATH:-$INSTALL_DIR/codecrafters}
 
 DOWNLOAD_URL="https://github.com/codecrafters-io/cli/releases/download/${VERSION}/${VERSION}_${OS}_${ARCH}.tar.gz"
 
 echo "This script will automatically install codecrafters (${VERSION}) for you."
-echo "You will be prompted for your password by sudo."
+echo "You will be prompted for your password by sudo if needed."
```
Comment (rohitpaulk) [inline install.sh:38]: Good idea to only `sudo` if needed!

File: install.sh (hunk 1 above)
Comment (rohitpaulk) [reply install.sh:71]: (I saw your note in the description re: `+x` being disabled in tmp, just want to confirm whether this is best/common practice)

File: install.sh (hunk 2 above)
Comment (nikandfor) [reply install.sh:61]: I've added this when I saw some progress bar, but didn't see what it is about. It's a bit scary to see something happen on your laptop, and it's continuous process, but you don't know what it does.

File: install.sh (hunk 1 above)
Comment (nikandfor) [reply install.sh:71]: Agree, this is probably not much compatible. Even flags are totally different. Maybe change the order of mv+chmod then?

File: install.sh (hunk 2 above)
Comment (rohitpaulk) [reply install.sh:61]: Ah, okay got it. Can we say "Downloading CodeCrafters CLI..." or something like that ("archive" isn't important from the user's perspective).

File: install.sh (hunk 1 above)
Comment (rohitpaulk) [reply install.sh:71]: @nikandfor Yep, that works!

Comment (nikandfor) [reply install.sh:71]: What is the purpose of -k option for sudo here?

Comment (rohitpaulk) [reply install.sh:71]: I just copied the `sudo -k` from one of the scripts linked to above - honestly had no clue what it does 🙂 I think it's fine to `chmod` in the tmp folder as you suggested. I'll keep an eye out for user complaints. Fwiw I haven't heard any complaints re: `chmod` failing yet, but we have heard a couple about posix sh compatibility already.

In my opinion, If there isn't an "obvious" answer to something, averaging approaches from the scripts linked above is a good strategy - these are used by 1000s of people, so there's some quality control already baked in.

**Comment by nikandfor** [comment]
Added test workflow which runs the script with the dash shell. Dash is a minimalist POSIX-compliant shell. So no bash or anything specific features should pass the test.

**Comment by rohitpaulk** [comment]
Excellent idea to add a test script 🙂

**Review by rohitpaulk — APPROVED** [review]

**Pull request merged**